### PR TITLE
Validate the freshness of session nonce

### PR DESCRIPTION
**Pull request by chendave**
According to the RMM spec, A7.2.3 Attestation token format [1], the CCA token has a client challenge sealed in the token, the challenge is generated when the CCA token is got from AA (attestation agent).

When the CCA token is validated by the veraison, a session nonce is created as well, semantically, the session nonce should be equal with the client challenge, this logic has been validated inside veraison service [2].

While the freshness also implies the session nonce key should equal with the nonce enveloped in the ear which is the format of attestation result [3], by this, we have:
```
client challenge in token == session nonce == ear's nonce
```
and thus avoid the replay attack.

NOTE: If the "client challenge in token != session nonce", vts service from Veraison takes it as a warning, this is debatable and should be further discuss with Veraison team, `rust-client` should return an error instead.

fix: https://github.com/confidential-containers/attestation-service/issues/127

> WARN    vts     {"detail":["freshness: realm challenge (00000000000...) does not match session nonce
> ... "detail-type":"error","error":"bad evidence"}

Signed-off-by: Dave Chen <dave.chen@arm.com>

[1] https://documentation-service.arm.com/static/63a16f163f28e5456434c719?token=
[2] https://github.com/veraison/services/blob/dfb068204473cad9c412337d5abef7ad88b8bc3b/scheme/cca-ssd-platform/evidence_handler.go#L120-L126
[3] https://github.com/veraison/docs/blob/main/architecture/verifier/freshness.md

### CODE REVIEW

File: attestation-service/Cargo.toml [hunk 1]
```diff
@@ -61,7 +61,9 @@ tokio.workspace = true
 tonic = { workspace = true, optional = true }
 uuid = { version = "1.1.2", features = ["v4"] }
 veraison-apiclient = { git = "https://github.com/chendave/rust-apiclient", branch = "token", optional = true }
+ear = { git = "https://github.com/chendave/rust-ear", branch = "test" }
```
Comment (chendave) [inline attestation-service/Cargo.toml:64]: need to sync with upstream once my change in the fork have been merged.

**Comment by chendave** [comment]
cc @Xynnn007 @jialez0

File: attestation-service/src/verifier/cca/mod.rs [hunk 2]
```diff
@@ -77,16 +85,42 @@ impl Verifier for CCA {
...
+
+        let ear_nonce = plain_ear.nonce.unwrap();
+        let nonce_byte = engine::GeneralPurpose::new(&alphabet::URL_SAFE, general_purpose::PAD)
+            .decode(ear_nonce.to_string())
+            .unwrap();
+
+        if hash_of_nonce_pubkey != nonce_byte {
+            return Err(anyhow!(
+                "HASH(nonce||pubkey) is different from that in ear's session nonce"
+            ));
```
Comment (Xynnn007) [inline attestation-service/src/verifier/cca/mod.rs:123]:
```suggestion
            bail!("HASH(nonce||pubkey) is different from that in ear's session nonce");
```

File: attestation-service/src/verifier/cca/mod.rs [hunk 3]
```diff
@@ -77,16 +85,42 @@ impl Verifier for CCA {
...
+
+        // Get back the pub key to decrypt the ear which holds raw evidence and the session nonce
+        let public_key_pem = verification_api.ear_verification_key_as_pem()?;
+        let dk = jwt::DecodingKey::from_ec_pem(public_key_pem.as_bytes()).unwrap();
+        let plain_ear = Ear::from_jwt(result.as_str(), jwt::Algorithm::ES256, &dk).unwrap();
+
+        let ear_nonce = plain_ear.nonce.unwrap();
+        let nonce_byte = engine::GeneralPurpose::new(&alphabet::URL_SAFE, general_purpose::PAD)
+            .decode(ear_nonce.to_string())
+            .unwrap();
```
Comment (Xynnn007) [inline attestation-service/src/verifier/cca/mod.rs:118]:
```suggestion
        let nonce_byte = base64::engine::general_purpose::URL_SAFE.decode(ear_nonce)
            .context("decode nonce byte from ear")?;
```

File: attestation-service/src/verifier/cca/mod.rs [hunk 4]
```diff
@@ -77,16 +85,42 @@ impl Verifier for CCA {
...
-        match cr.run(nonce, my_evidence_builder, token.clone()).await {
+        let n = Nonce::Value(hash_of_nonce_pubkey.clone());
+        let result = match cr.run(n, my_evidence_builder, token.clone()).await {
             Err(e) => {
                 log::error!("Error: {}", e);
-                return Err(anyhow!("Attestation failed with error: {:?}", e));
-            }
-            Ok(attestation_result) => {
-                log::info!("Attestation Result: {}", attestation_result)
+                panic!("Attestation failed with error: {:?}", e);
```
Comment (Xynnn007) [inline attestation-service/src/verifier/cca/mod.rs:105]: Let's try not to use `panic` or `unwrap` inside the codes.
```suggestion
                bail!("CCA Attestation failed with error: {:?}", e);
```

File: attestation-service/Cargo.toml [hunk 5]
```diff
@@ -61,7 +61,9 @@ tokio.workspace = true
...
 x509-parser = { version = "0.14.0", optional = true }
+jsonwebtoken = "8"
```
Comment (Xynnn007) [inline attestation-service/Cargo.toml:66]: Could you help to reorder the new two deps in dictionarial order?

**Comment by chendave** [comment]
@Xynnn007  lint fix is here: https://github.com/confidential-containers/attestation-service/pull/148

File: attestation-service/src/verifier/cca/mod.rs (hunk 2 above)
Comment (chendave) [reply attestation-service/src/verifier/cca/mod.rs]: Done

File: attestation-service/src/verifier/cca/mod.rs (hunk 3 above)
Comment (chendave) [reply attestation-service/src/verifier/cca/mod.rs]: Done

File: attestation-service/src/verifier/cca/mod.rs (hunk 4 above)
Comment (chendave) [reply attestation-service/src/verifier/cca/mod.rs]: Done

File: attestation-service/Cargo.toml (hunk 5 above)
Comment (chendave) [reply attestation-service/Cargo.toml]: Done

File: attestation-service/src/verifier/cca/mod.rs [hunk 6]
```diff
@@ -77,16 +81,40 @@ impl Verifier for CCA {
...
             }
+            Ok(attestation_result) => attestation_result,
+        };
+
+        // Get back the pub key to decrypt the ear which holds raw evidence and the session nonce
+        let public_key_pem = verification_api.ear_verification_key_as_pem()?;
+        let dk = jwt::DecodingKey::from_ec_pem(public_key_pem.as_bytes()).unwrap();
+        let plain_ear = Ear::from_jwt(result.as_str(), jwt::Algorithm::ES256, &dk).unwrap();
+
+        let ear_nonce = plain_ear.nonce.unwrap();
```
Comment (Xynnn007) [inline attestation-service/src/verifier/cca/mod.rs:111]: Please try to replace `.unwrap()`s with error handling, s.t. `.context("...")?`

File: attestation-service/Cargo.toml (hunk 1 above)
Comment (chendave) [reply attestation-service/Cargo.toml:65]: two of them still pending: https://github.com/veraison/rust-ear/pulls?q=is%3Apr+author%3Achendave+

File: attestation-service/src/verifier/cca/mod.rs (hunk 6 above)
Comment (chendave) [reply attestation-service/src/verifier/cca/mod.rs]: Done

File: attestation-service/src/verifier/cca/mod.rs [hunk 7]
```diff
@@ -77,16 +81,42 @@ impl Verifier for CCA {
...
 
-        let nonce = Nonce::Size(32);
+        let mut hasher = Sha384::new();
+        hasher.update(&nonce);
+        hasher.update(&attestation.tee_pubkey.k_mod);
+        hasher.update(&attestation.tee_pubkey.k_exp);
+        let mut hash_of_nonce_pubkey = hasher.finalize().to_vec();
+        hash_of_nonce_pubkey.resize(64, 0);
+
+        log::info!(
```
Comment (portersrc) [inline attestation-service/src/verifier/cca/mod.rs:91]: Consider `debug!` for this?

File: attestation-service/Cargo.toml (hunk 1 above)
Comment (chendave) [reply attestation-service/Cargo.toml]: @Xynnn007 this was updated.

**Comment by chendave** [comment]
@jialez0 @Xynnn007 any more comments for this?

**Review by jialez0 — APPROVED** [review]

**Pull request merged**